waiter/src/waiter/scheduler/marathon.clj

```diff
@@ -395,11 +419,143 @@
   (service-id->state [_ service-id]
     {:failed-instances (service-id->failed-instances service-id->failed-instances-transient-store service-id)
      :killed-instances (scheduler/service-id->killed-instances service-id)
-     :kill-info (get @service-id->kill-info-store service-id)})
+     :kill-info (get @service-id->kill-info-store service-id)
+     :out-of-sync-state (get @service-id->out-of-sync-state-store service-id)})
 
   (state [_]
     {:service-id->failed-instances-transient-store @service-id->failed-instances-transient-store
-     :service-id->kill-info-store @service-id->kill-info-store}))
+     :service-id->kill-info-store @service-id->kill-info-store
+     :service-id->out-of-sync-state-store @service-id->out-of-sync-state-store}))
+
+(defn- get-apps-with-deployments
+  "Retrieves the apps with the deployment info embedded."
+  [{:keys [marathon-api is-waiter-app?-fn]}]
+  (get-apps marathon-api is-waiter-app?-fn {"embed" ["apps.deployments" "apps.tasks"]}))
+
+(defn retrieve-service-id->out-of-sync-state
+  "Retrieves the service-id->out-of-sync-state."
+  [{:keys [service-id->out-of-sync-state-store]}]
+  (deref service-id->out-of-sync-state-store))
+
+(defn- reset-service-id->out-of-sync-state!
+  "Updates the service-id->out-of-sync-state."
+  [{:keys [service-id->out-of-sync-state-store]} service-id->out-of-sync-state]
+  (reset! service-id->out-of-sync-state-store service-id->out-of-sync-state))
+
+(defn- is-out-of-sync?
+  "Returns true if the service does not have pending deployments and the instance and task counts do not match."
+  [{:keys [deployments instances tasks]}]
+  (and (empty? deployments) (not= instances (count tasks))))
+
+(defn- retrieve-out-of-sync-apps
+  "Retrieves the details about out-of-sync services.
+   Returns a map keyed with service-id and values of :instances-requested and :instances-scheduled."
+  [marathon-scheduler]
+  (try
+    (->> (get-apps-with-deployments marathon-scheduler)
+         (filter is-out-of-sync?)
+         (map (fn [{:keys [id instances tasks]}]
+                [(remove-slash-prefix id)
+                 {:instances-requested instances :instances-scheduled (count tasks)}]))
+         (into {}))
+    (catch Exception e
+      (log/error e "unable to retrieve out-of-sync services"))))
+
+(defn- trigger-sync-deployment!
+  "Triggers deployment for out-of-sync service."
+  [marathon-scheduler service-id {:keys [instances-scheduled] :as task-data}]
+  (try
+    (log/info "triggering sync deployment" {:service-id service-id :task-data task-data})
+    (scheduler/scale-app marathon-scheduler service-id instances-scheduled false)
+    (catch Exception e
+      (log/error e "unable to sync marathon deployment for" service-id))))
+
+(defn- process-out-of-sync-services!
```

Comment:
This function could be split into 2 parts:

f(out-of-sync-state) -> (out-of-sync-state', trigger-service-ids)
f!(trigger-service-ids)

I believe it makes sense to do so.